src/libraries/System.Private.Uri/tests/FunctionalTests/UriEscapingTest.cs

```diff
@@ -24,193 +24,304 @@ public class UriEscapingTest
-        {
-            string input = RFC2396Unreserved;
-            string output = Uri.UnescapeDataString(Escape(input));
-            Assert.Equal(input, output);
-        }
+                if (input == expectedOutput)
+                {
+                    Assert.Same(input, output);
+                }
 
-        [Fact]
-        public void UriUnescapingDataString_RFC2396Reserved_Unescaped()
-        {
-            string input = RFC2396Reserved;
-            string output = Uri.UnescapeDataString(Escape(input));
-            Assert.Equal(input, output);
-        }
+                // Span overload
+                output = Uri.EscapeDataString(input.AsSpan());
+                Assert.Equal(expectedOutput, output);
 
-        [Fact]
-        public void UriUnescapingDataString_RFC3986Unreserved_Unescaped()
-        {
-            string input = RFC3986Unreserved;
-            string output = Uri.UnescapeDataString(Escape(input));
-            Assert.Equal(input, output);
-        }
+                char[] destination = new char[expectedOutput.Length + 2];
 
-        [Fact]
-        public void UriUnescapingDataString_RFC3986Reserved_Unescaped()
-        {
-            string input = RFC3986Reserved;
-            string output = Uri.UnescapeDataString(Escape(input));
-            Assert.Equal(input, output);
-        }
+                // Exact destination size
+                Assert.True(Uri.TryEscapeDataString(input, destination.AsSpan(0, expectedOutput.Length), out int charsWritten));
+                Assert.Equal(expectedOutput.Length, charsWritten);
+                Assert.Equal(expectedOutput, destination.AsSpan(0, charsWritten));
 
-        [Fact]
-        public void UriUnescapingDataString_RFC3986ReservedWithIRI_Unescaped()
-        {
-            // Note that \ and % are not officialy reserved, but we treat it as reserved.
-            string input = RFC3986Reserved;
-            string output = Uri.UnescapeDataString(Escape(input));
-            Assert.Equal(input, output);
-        }
+                // Larger destination
+                Assert.True(Uri.TryEscapeDataString(input, destination.AsSpan(1), out charsWritten));
+                Assert.Equal(expectedOutput.Length, charsWritten);
+                Assert.Equal(expectedOutput, destination.AsSpan(1, charsWritten));
 
-        [Fact]
-        public void UriUnescapingDataString_Unicode_Unescaped()
-        {
-            string input = @"\u30AF";
-            string output = Uri.UnescapeDataString(Escape(input));
-            Assert.Equal(input, output);
+                // Destination too small
+                if (expectedOutput.Length > 0)
+                {
+                    Assert.False(Uri.TryEscapeDataString(input, destination.AsSpan(0, expectedOutput.Length - 1), out charsWritten));
+                    Assert.Equal(0, charsWritten);
+                }
+
+                // Overlapped source/destination
```

Comment:
Supporting overlapped source/destinations if "free" with the current implementation approach (via `ValueStringBuilder`).
Alternatively, we could detect and throw here if we wanted to.